File: schemas/analytics-metrics/convex/event_properties.ts
// event_properties: Key-value properties for events.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const event_properties = defineTable({
  eventId: v.id("events"),
  key: v.string(),
  value: v.string(),
})
  .index("by_event_id_and_key", ["eventId", "key"])
  .index("by_key_and_value", ["key", "value"]);

File: schemas/analytics-metrics/spacetimedb/event_properties.rs
// event_properties: Key-value property pairs attached to individual events.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

// Composite unique constraint (not expressible inline):
// - unique(event_id, key)

#[spacetimedb::table(name = event_properties, public)]
pub struct EventProperty {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub event_id: String, // UUID, FK → events.id (cascade delete)

    // Composite index: (key, value) — enforce in reducer logic
    pub key: String,
    pub value: String,
    pub created_at: Timestamp,
}
